src/utils.py

```diff
@@ -797,303 +832,157 @@ def ops_and_specs_to_dummy_nsrts(strips_ops: Sequence[STRIPSOperator],
-            # Check whether we already expanded this fact.
-            if not fact.expanded:
-                # Iterate over all operators this fact is a precondition of.
-                for operator in fact.precondition_of:
-                    # Decrease the precondition counter.
-                    operator.counter -= 1
-                    # Check whether all preconditions are True and we can apply
-                    # this operator.
-                    if operator.counter <= 0:
-                        for n in operator.add_effects:
-                            neighbor = self.facts[n]
-                            # Calculate the cost of applying this operator.
-                            tmp_dist = self.get_cost(operator)
-                            if tmp_dist < neighbor.distance:
-                                # If the new costs are cheaper than the old
-                                # costs, we change the neighbor's heuristic
-                                # values.
-                                neighbor.distance = tmp_dist
-                                neighbor.cheapest_achiever = operator
-                                # And push it on the queue.
-                                hq.heappush(queue, (
-                                    tmp_dist, self.tie_breaker, neighbor))
-                                self.tie_breaker += 1
-                # Finally the fact is marked as expanded.
-                fact.expanded = True
-
-
-class _HAddHeuristic(_RelaxationHeuristic):
-    """Implements the HAdd delete relaxation heuristic.
+    @functools.lru_cache(maxsize=None)
+    def _evaluate(pyperplan_facts: _PyperplanFacts,
+                  pyperplan_goal: _PyperplanFacts,
+                  pyperplan_heuristic: _PyperplanBaseHeuristic) -> float:
+        pyperplan_node = _PyperplanNode(pyperplan_facts, pyperplan_goal)
+        return pyperplan_heuristic(pyperplan_node)
+
+
+def _create_pyperplan_task(init_atoms: Set[GroundAtom],
+                           goal: Set[GroundAtom],
+                           ground_ops: Collection[GroundNSRTOrSTRIPSOperator],
+                           predicates: Collection[Predicate],
+                           objects: Collection[Object],
+                           static_atoms: Set[GroundAtom],
+                           ) -> _PyperplanTask:
+    """Helper glue for pyperplan heuristics.
     """
-    @staticmethod
-    def _accumulate(distances: Collection[float]) -> float:
-        return sum(distances)
+    all_atoms = get_all_ground_atoms(frozenset(predicates), frozenset(objects))
+    # Note: removing static atoms.
+    pyperplan_facts = _atoms_to_tuples(all_atoms - static_atoms)
+    pyperplan_state = _atoms_to_tuples(init_atoms - static_atoms)
+    pyperplan_goal = _atoms_to_tuples(goal - static_atoms)
+    pyperplan_operators = set()
+    for op in ground_ops:
+        # Note: the pyperplan operator must include the objects, because hFF
```

Comment:
I don't understand why the "because..." implies this note.

Reply:
I'll expand the comment:
```
        # Note: the pyperplan operator must include the objects, because hFF
        # uses the operator name in constructing the relaxed plan, and the
        # relaxed plan is a set. If we instead just used op.name, there would
        # be a very nasty bug where two ground operators in the relaxed plan
        # that have different objects are counted as just one.
```

Reply:
no idea how you figured that out, great stuff!